File: src/org/kde/kdeconnect/Plugins/MprisPlugin/MprisNowPlayingFragment.java
package org.kde.kdeconnect.Plugins.MprisPlugin;

import android.content.ActivityNotFoundException;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.graphics.Bitmap;
import android.graphics.drawable.Drawable;
import android.net.Uri;
import android.os.Bundle;
import android.os.Handler;
import android.preference.PreferenceManager;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.SeekBar;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.content.ContextCompat;
import androidx.core.graphics.drawable.DrawableCompat;
import androidx.fragment.app.Fragment;

import org.apache.commons.lang3.ArrayUtils;
import org.kde.kdeconnect.Backends.BaseLink;
import org.kde.kdeconnect.Backends.BaseLinkProvider;
import org.kde.kdeconnect.BackgroundService;
import org.kde.kdeconnect.Helpers.VideoUrlsHelper;
import org.kde.kdeconnect.Helpers.VolumeHelperKt;
import org.kde.kdeconnect.NetworkPacket;
import org.kde.kdeconnect_tp.R;
import org.kde.kdeconnect_tp.databinding.MprisControlBinding;
import org.kde.kdeconnect_tp.databinding.MprisNowPlayingBinding;

import java.net.MalformedURLException;
import java.util.List;

public class MprisNowPlayingFragment extends Fragment implements VolumeKeyListener {

    final static int MENU_OPEN_URL = Menu.FIRST;
    private final Handler positionSeekUpdateHandler = new Handler();
    MprisControlBinding mprisControlBinding;
    private MprisNowPlayingBinding activityMprisBinding;
    private String deviceId;
    private Runnable positionSeekUpdateRunnable = null;
    private MprisPlugin.MprisPlayer targetPlayer = null;
    private final BaseLinkProvider.ConnectionReceiver connectionReceiver = new BaseLinkProvider.ConnectionReceiver() {
        @Override
        public void onConnectionReceived(NetworkPacket identityPacket, BaseLink link) {
            connectToPlugin(null);
        }

        @Override
        public void onConnectionLost(BaseLink link) {

        }
    };

    public static MprisNowPlayingFragment newInstance(String deviceId) {
        MprisNowPlayingFragment mprisNowPlayingFragment = new MprisNowPlayingFragment();

        Bundle arguments = new Bundle();
        arguments.putString(MprisPlugin.DEVICE_ID_KEY, deviceId);

        mprisNowPlayingFragment.setArguments(arguments);

        return mprisNowPlayingFragment;
    }

    private static String milisToProgress(long milis) {
        int length = (int) (milis / 1000); //From milis to seconds
        StringBuilder text = new StringBuilder();
        int minutes = length / 60;
        if (minutes > 60) {
            int hours = minutes / 60;
            minutes = minutes % 60;
            text.append(hours).append(':');
            if (minutes < 10) text.append('0');
        }
        text.append(minutes).append(':');
        int seconds = (length % 60);
        if (seconds < 10)
            text.append('0'); // needed to show length properly (eg 4:05 instead of 4:5)
        text.append(seconds);
        return text.toString();
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {

        if (activityMprisBinding == null) {
            activityMprisBinding = MprisNowPlayingBinding.inflate(inflater);
            mprisControlBinding = activityMprisBinding.mprisControl;

            String targetPlayerName = "";
            Intent activityIntent = requireActivity().getIntent();
            activityIntent.getStringExtra("player");
            activityIntent.removeExtra("player");

            if (TextUtils.isEmpty(targetPlayerName)) {
                if (savedInstanceState != null) {
                    targetPlayerName = savedInstanceState.getString("targetPlayer");
                }
            }

            deviceId = requireArguments().getString(MprisPlugin.DEVICE_ID_KEY);

            SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(requireContext());
            String interval_time_str = prefs.getString(getString(R.string.mpris_time_key),
                    getString(R.string.mpris_time_default));
            final int interval_time = Integer.parseInt(interval_time_str);

            BackgroundService.RunCommand(requireContext(), service -> service.addConnectionListener(connectionReceiver));
            connectToPlugin(targetPlayerName);

            performActionOnClick(mprisControlBinding.loopButton, p -> {
                switch (p.getLoopStatus()) {
                    case "None":
                        p.setLoopStatus("Track");
                        break;
                    case "Track":
                        p.setLoopStatus("Playlist");
                        break;
                    case "Playlist":
                        p.setLoopStatus("None");
                        break;
                }
            });

            performActionOnClick(mprisControlBinding.playButton, MprisPlugin.MprisPlayer::playPause);

            performActionOnClick(mprisControlBinding.shuffleButton, p -> p.setShuffle(!p.getShuffle()));

            performActionOnClick(mprisControlBinding.prevButton, MprisPlugin.MprisPlayer::previous);

            performActionOnClick(mprisControlBinding.rewButton, p -> targetPlayer.seek(interval_time * -1));

            performActionOnClick(mprisControlBinding.ffButton, p -> p.seek(interval_time));

            performActionOnClick(mprisControlBinding.nextButton, MprisPlugin.MprisPlayer::next);

            performActionOnClick(mprisControlBinding.stopButton, MprisPlugin.MprisPlayer::stop);

            mprisControlBinding.volumeSeek.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {
                @Override
                public void onProgressChanged(SeekBar seekBar, int i, boolean b) {
                }

                @Override
                public void onStartTrackingTouch(SeekBar seekBar) {
                }

                @Override
                public void onStopTrackingTouch(final SeekBar seekBar) {
                    BackgroundService.RunCommand(requireContext(), service -> {
                        if (targetPlayer == null) return;
                        targetPlayer.setVolume(seekBar.getProgress());
                    });
                }
            });

            positionSeekUpdateRunnable = () -> {
                Context context = getContext();
                if (context == null) return; // Fragment was already detached
                BackgroundService.RunCommand(context, service -> {
                    if (targetPlayer != null) {
                        mprisControlBinding.positionSeek.setProgress((int) (targetPlayer.getPosition()));
                    }
                    positionSeekUpdateHandler.removeCallbacks(positionSeekUpdateRunnable);
                    positionSeekUpdateHandler.postDelayed(positionSeekUpdateRunnable, 1000);
                });
            };
            positionSeekUpdateHandler.postDelayed(positionSeekUpdateRunnable, 200);

            mprisControlBinding.positionSeek.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {
                @Override
                public void onProgressChanged(SeekBar seekBar, int progress, boolean byUser) {
                    mprisControlBinding.progressTextview.setText(milisToProgress(progress));
                }

                @Override
                public void onStartTrackingTouch(SeekBar seekBar) {
                    positionSeekUpdateHandler.removeCallbacks(positionSeekUpdateRunnable);
                }

                @Override
                public void onStopTrackingTouch(final SeekBar seekBar) {
                    BackgroundService.RunCommand(requireContext(), service -> {
                        if (targetPlayer != null) {
                            targetPlayer.setPosition(seekBar.getProgress());
                        }
                        positionSeekUpdateHandler.postDelayed(positionSeekUpdateRunnable, 200);
                    });
                }
            });

            mprisControlBinding.nowPlayingTextview.setSelected(true);

        }

        return activityMprisBinding.getRoot();
    }

    private void connectToPlugin(final String targetPlayerName) {
        BackgroundService.RunWithPlugin(requireContext(), deviceId, MprisPlugin.class, mpris -> {
            targetPlayer = mpris.getPlayerStatus(targetPlayerName);

            mpris.setPlayerStatusUpdatedHandler("activity", () -> requireActivity().runOnUiThread(() -> updatePlayerStatus(mpris)));
            mpris.setPlayerListUpdatedHandler("activity", () -> {
                final List<String> playerList = mpris.getPlayerList();
                final ArrayAdapter<String> adapter = new ArrayAdapter<>(requireContext(),
                        android.R.layout.simple_spinner_item,
                        playerList.toArray(ArrayUtils.EMPTY_STRING_ARRAY)
                );

                adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
                requireActivity().runOnUiThread(() -> {
                    mprisControlBinding.playerSpinner.setAdapter(adapter);

                    if (playerList.isEmpty()) {
                        mprisControlBinding.noPlayers.setVisibility(View.VISIBLE);
                        mprisControlBinding.playerSpinner.setVisibility(View.GONE);
                        mprisControlBinding.nowPlayingTextview.setText("");
                    } else {
                        mprisControlBinding.noPlayers.setVisibility(View.GONE);
                        mprisControlBinding.playerSpinner.setVisibility(View.VISIBLE);
                    }

                    mprisControlBinding.playerSpinner.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
                        @Override
                        public void onItemSelected(AdapterView<?> arg0, View arg1, int pos, long id) {

                            if (pos >= playerList.size()) return;

                            String player = playerList.get(pos);
                            if (targetPlayer != null && player.equals(targetPlayer.getPlayer())) {
                                return; //Player hasn't actually changed
                            }
                            targetPlayer = mpris.getPlayerStatus(player);
                            updatePlayerStatus(mpris);

                            if (targetPlayer != null && targetPlayer.isPlaying()) {
                                MprisMediaSession.getInstance().playerSelected(targetPlayer);
                            }
                        }

                        @Override
                        public void onNothingSelected(AdapterView<?> arg0) {
                            targetPlayer = null;
                        }
                    });

                    if (targetPlayer == null) {
                        //If no player is selected, try to select a playing player
                        targetPlayer = mpris.getPlayingPlayer();
                    }
                    //Try to select the specified player
                    if (targetPlayer != null) {
                        int targetIndex = adapter.getPosition(targetPlayer.getPlayer());
                        if (targetIndex >= 0) {
                            mprisControlBinding.playerSpinner.setSelection(targetIndex);
                        } else {
                            targetPlayer = null;
                        }
                    }
                    //If no player selected, select the first one (if any)
                    if (targetPlayer == null && !playerList.isEmpty()) {
                        targetPlayer = mpris.getPlayerStatus(playerList.get(0));
                        mprisControlBinding.playerSpinner.setSelection(0);
                    }
                    updatePlayerStatus(mpris);
                });
            });
        });
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        BackgroundService.RunCommand(requireContext(), service -> service.removeConnectionListener(connectionReceiver));
    }

    private void performActionOnClick(View v, MprisPlayerCallback l) {
        v.setOnClickListener(view -> BackgroundService.RunCommand(requireContext(), service -> {
            if (targetPlayer == null) return;
            l.performAction(targetPlayer);
        }));
    }

    private void updatePlayerStatus(MprisPlugin mpris) {
        MprisPlugin.MprisPlayer playerStatus = targetPlayer;
        if (playerStatus == null) {
            //No player with that name found, just display "empty" data
            playerStatus = mpris.getEmptyPlayer();
        }
        String song = playerStatus.getCurrentSong();

        if (!mprisControlBinding.nowPlayingTextview.getText().toString().equals(song)) {
            mprisControlBinding.nowPlayingTextview.setText(song);
        }

        Bitmap albumArt = playerStatus.getAlbumArt();
        if (albumArt == null) {
            final Drawable drawable = ContextCompat.getDrawable(requireContext(), R.drawable.ic_album_art_placeholder);
            assert drawable != null;
            Drawable placeholder_art = DrawableCompat.wrap(drawable);
            DrawableCompat.setTint(placeholder_art, ContextCompat.getColor(requireContext(), R.color.primary));
            activityMprisBinding.albumArt.setImageDrawable(placeholder_art);
        } else {
            activityMprisBinding.albumArt.setImageBitmap(albumArt);
        }

        if (playerStatus.isSeekAllowed()) {
            mprisControlBinding.timeTextview.setText(milisToProgress(playerStatus.getLength()));
            mprisControlBinding.positionSeek.setMax((int) (playerStatus.getLength()));
            mprisControlBinding.positionSeek.setProgress((int) (playerStatus.getPosition()));
            mprisControlBinding.progressSlider.setVisibility(View.VISIBLE);
        } else {
            mprisControlBinding.progressSlider.setVisibility(View.GONE);
        }

        int volume = playerStatus.getVolume();
        mprisControlBinding.volumeSeek.setProgress(volume);
        if(!playerStatus.isSetVolumeAllowed()) {
            mprisControlBinding.volumeSeek.setEnabled(false);
        }
        boolean isPlaying = playerStatus.isPlaying();
        if (isPlaying) {
            mprisControlBinding.playButton.setIconResource(R.drawable.ic_pause_black);
            mprisControlBinding.playButton.setEnabled(playerStatus.isPauseAllowed());
        } else {
            mprisControlBinding.playButton.setIconResource(R.drawable.ic_play_black);
            mprisControlBinding.playButton.setEnabled(playerStatus.isPlayAllowed());
        }

        String loopStatus = playerStatus.getLoopStatus();
        switch (loopStatus) {
            case "None":
                mprisControlBinding.loopButton.setIconResource(R.drawable.ic_loop_none_black);
                break;
            case "Track":
                mprisControlBinding.loopButton.setIconResource(R.drawable.ic_loop_track_black);
                break;
            case "Playlist":
                mprisControlBinding.loopButton.setIconResource(R.drawable.ic_loop_playlist_black);
                break;
        }

        boolean shuffle = playerStatus.getShuffle();
        if (shuffle) {
            mprisControlBinding.shuffleButton.setIconResource(R.drawable.ic_shuffle_on_black);
        } else {
            mprisControlBinding.shuffleButton.setIconResource(R.drawable.ic_shuffle_off_black);
        }

        mprisControlBinding.loopButton.setVisibility(playerStatus.isLoopStatusAllowed() ? View.VISIBLE : View.GONE);
        mprisControlBinding.shuffleButton.setVisibility(playerStatus.isShuffleAllowed() ? View.VISIBLE : View.GONE);
        mprisControlBinding.volumeLayout.setVisibility(playerStatus.isSetVolumeAllowed() ? View.VISIBLE : View.GONE);
        mprisControlBinding.rewButton.setVisibility(playerStatus.isSeekAllowed() ? View.VISIBLE : View.GONE);
        mprisControlBinding.ffButton.setVisibility(playerStatus.isSeekAllowed() ? View.VISIBLE : View.GONE);

        requireActivity().invalidateOptionsMenu();

        //Show and hide previous/next buttons simultaneously
        if (playerStatus.isGoPreviousAllowed() || playerStatus.isGoNextAllowed()) {
            mprisControlBinding.prevButton.setVisibility(View.VISIBLE);
            mprisControlBinding.prevButton.setEnabled(playerStatus.isGoPreviousAllowed());
            mprisControlBinding.nextButton.setVisibility(View.VISIBLE);
            mprisControlBinding.nextButton.setEnabled(playerStatus.isGoNextAllowed());
        } else {
            mprisControlBinding.prevButton.setVisibility(View.GONE);
            mprisControlBinding.nextButton.setVisibility(View.GONE);
        }
    }

    /**
     * Change current volume with provided step.
     *
     * @param step step size volume change
     */
    private void updateVolume(int step) {
        if (targetPlayer == null) return;

        int newVolume = VolumeHelperKt.calculateNewVolume(targetPlayer.getVolume(), VolumeHelperKt.DEFAULT_MAX_VOLUME, step);

        if (targetPlayer.getVolume() != newVolume) {
            targetPlayer.setVolume(newVolume);
        }
    }

    @Override
    public void onVolumeUp() {
        updateVolume(VolumeHelperKt.DEFAULT_VOLUME_STEP);
    }

    @Override
    public void onVolumeDown() {
        updateVolume(-VolumeHelperKt.DEFAULT_VOLUME_STEP);
    }

    @Override
    public void onPrepareOptionsMenu(@NonNull Menu menu) {
        menu.clear();
        if (targetPlayer != null && !"".equals(targetPlayer.getUrl())) {
            menu.add(0, MENU_OPEN_URL, Menu.NONE, R.string.mpris_open_url);
        }
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        if (targetPlayer != null && item.getItemId() == MENU_OPEN_URL) {
            try {
                String url = VideoUrlsHelper.formatUriWithSeek(targetPlayer.getUrl(), targetPlayer.getPosition()).toString();
                Intent browserIntent = new Intent(Intent.ACTION_VIEW, Uri.parse(url));
                startActivity(browserIntent);
                targetPlayer.pause();
                return true;
            } catch (MalformedURLException | ActivityNotFoundException e) {
                e.printStackTrace();
                Toast.makeText(requireContext(), getString(R.string.cant_open_url), Toast.LENGTH_LONG).show();
            }
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public void onSaveInstanceState(@NonNull Bundle outState) {
        if (targetPlayer != null) {
            outState.putString("targetPlayer", targetPlayer.getPlayer());
        }
    }

    private interface MprisPlayerCallback {
        void performAction(MprisPlugin.MprisPlayer player);
    }
}

File: src/org/kde/kdeconnect/Plugins/MprisPlugin/MprisNowPlayingFragment.kt
/*
 * SPDX-FileCopyrightText: 2023 Albert Vaca Cintora <albertvaka@gmail.com>
 *
 * SPDX-License-Identifier: GPL-2.0-only OR GPL-3.0-only OR LicenseRef-KDE-Accepted-GPL
*/
package org.kde.kdeconnect.Plugins.MprisPlugin

import android.content.ActivityNotFoundException
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.os.Handler
import android.preference.PreferenceManager
import android.util.Log
import android.view.*
import android.widget.AdapterView
import android.widget.ArrayAdapter
import android.widget.SeekBar
import android.widget.SeekBar.OnSeekBarChangeListener
import android.widget.Toast
import androidx.core.content.ContextCompat
import androidx.core.graphics.drawable.DrawableCompat
import androidx.fragment.app.Fragment
import org.apache.commons.lang3.StringUtils
import org.kde.kdeconnect.Helpers.DEFAULT_MAX_VOLUME
import org.kde.kdeconnect.Helpers.DEFAULT_VOLUME_STEP
import org.kde.kdeconnect.Helpers.VideoUrlsHelper
import org.kde.kdeconnect.Helpers.calculateNewVolume
import org.kde.kdeconnect.KdeConnect
import org.kde.kdeconnect.Plugins.MprisPlugin.MprisPlugin.MprisPlayer
import org.kde.kdeconnect_tp.R
import org.kde.kdeconnect_tp.databinding.MprisControlBinding
import org.kde.kdeconnect_tp.databinding.MprisNowPlayingBinding
import java.net.MalformedURLException
import kotlin.time.Duration
import kotlin.time.Duration.Companion.milliseconds

private typealias MprisPlayerCallback = (MprisPlayer) -> Unit

class MprisNowPlayingFragment : Fragment(), VolumeKeyListener {
    private val positionSeekUpdateHandler = Handler()
    private lateinit var mprisControlBinding: MprisControlBinding
    private lateinit var activityMprisBinding: MprisNowPlayingBinding
    private var deviceId: String? = null
    private lateinit var positionSeekUpdateRunnable: Runnable

    private var targetPlayerName = ""
    private var targetPlayer: MprisPlayer? = null

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        activityMprisBinding = MprisNowPlayingBinding.inflate(inflater)
        mprisControlBinding = activityMprisBinding.mprisControl

        deviceId = requireArguments().getString(MprisPlugin.DEVICE_ID_KEY)

        val activityIntent = requireActivity().intent

        targetPlayerName = if (activityIntent.hasExtra("player")) {
            activityIntent.getStringExtra("player")!!.also {
                activityIntent.removeExtra("player")
            }
        } else {
            savedInstanceState?.getString("targetPlayer") ?: "".also {
                Log.i("MprisNowPlayingFragment", "No `targetPlayer` specified in savedInstanceState")
            }
        }

        connectToPlugin()

        val prefs = PreferenceManager.getDefaultSharedPreferences(requireContext())
        val interval_time_str = prefs.getString(
            getString(R.string.mpris_time_key),
            getString(R.string.mpris_time_default)
        )
        val interval_time = interval_time_str!!.toInt()

        performActionOnClick(mprisControlBinding.loopButton) { p: MprisPlayer ->
            when (p.loopStatus) {
                "None" -> p.sendSetLoopStatus("Track")
                "Track" -> p.sendSetLoopStatus("Playlist")
                "Playlist" -> p.sendSetLoopStatus("None")
            }
        }

        performActionOnClick(mprisControlBinding.playButton, MprisPlayer::sendPlayPause)

        performActionOnClick(
            mprisControlBinding.shuffleButton
        ) { p -> p.sendSetShuffle(!p.shuffle) }

        performActionOnClick(mprisControlBinding.prevButton, MprisPlayer::sendPrevious)

        performActionOnClick(
            mprisControlBinding.rewButton
        ) { p -> p.sendSeek(interval_time * -1) }

        performActionOnClick(
            mprisControlBinding.ffButton
        ) { p -> p.sendSeek(interval_time) }

        performActionOnClick(mprisControlBinding.nextButton, MprisPlayer::sendNext)

        performActionOnClick(mprisControlBinding.stopButton, MprisPlayer::sendStop)

        mprisControlBinding.volumeSeek.setOnSeekBarChangeListener(object : OnSeekBarChangeListener {
            override fun onProgressChanged(seekBar: SeekBar, i: Int, b: Boolean) {}

            override fun onStartTrackingTouch(seekBar: SeekBar) {}

            override fun onStopTrackingTouch(seekBar: SeekBar) {
                val targetPlayer = targetPlayer ?: return
                targetPlayer.sendSetVolume(seekBar.progress)
            }
        })

        positionSeekUpdateRunnable = Runnable {
            if (!isAdded) return@Runnable  // Fragment was already detached

            if (targetPlayer != null) {
                mprisControlBinding.positionSeek.progress = targetPlayer!!.position.toInt()
            }
            positionSeekUpdateHandler.removeCallbacks(positionSeekUpdateRunnable)
            positionSeekUpdateHandler.postDelayed(positionSeekUpdateRunnable, 1000)
        }
        positionSeekUpdateHandler.postDelayed(positionSeekUpdateRunnable, 200)

        mprisControlBinding.positionSeek.setOnSeekBarChangeListener(object : OnSeekBarChangeListener {
            override fun onProgressChanged(seekBar: SeekBar, progress: Int, byUser: Boolean) {
                mprisControlBinding.progressTextview.text = durationToProgress(progress.milliseconds)
            }

            override fun onStartTrackingTouch(seekBar: SeekBar) {
                positionSeekUpdateHandler.removeCallbacks(positionSeekUpdateRunnable)
            }

            override fun onStopTrackingTouch(seekBar: SeekBar) {
                if (targetPlayer != null) {
                    targetPlayer!!.sendSetPosition(seekBar.progress)
                }
                positionSeekUpdateHandler.postDelayed(positionSeekUpdateRunnable, 200)
            }
        })

        mprisControlBinding.nowPlayingTextview.isSelected = true

        return activityMprisBinding.root
    }

    override fun onDestroyView() {
        disconnectFromPlugin()
        super.onDestroyView()
    }

    private fun disconnectFromPlugin() {
        val plugin = KdeConnect.getInstance().getDevicePlugin(deviceId, MprisPlugin::class.java) ?: return
        plugin.apply {
            removePlayerListUpdatedHandler("activity")
            removePlayerStatusUpdatedHandler("activity")
        }
    }

    private fun connectToPlugin() {
        val plugin = KdeConnect.getInstance().getDevicePlugin(deviceId, MprisPlugin::class.java)
        if (plugin == null) {
            if (isAdded) {
                requireActivity().finish()
            }
            return
        }
        targetPlayer = plugin.getPlayerStatus(targetPlayerName)

        plugin.setPlayerStatusUpdatedHandler("activity") {
            requireActivity().runOnUiThread {
                updatePlayerStatus(plugin)
            }
        }
        plugin.setPlayerListUpdatedHandler("activity") {
            requireActivity().runOnUiThread {
                val playerList = plugin.playerList
                val adapter = ArrayAdapter(
                    requireContext(),
                    android.R.layout.simple_spinner_item,
                    playerList.toTypedArray()
                )

                adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)

                mprisControlBinding.playerSpinner.adapter = adapter

                if (playerList.isEmpty()) {
                    mprisControlBinding.noPlayers.visibility = View.VISIBLE
                    mprisControlBinding.playerSpinner.visibility = View.GONE
                    mprisControlBinding.nowPlayingTextview.text = ""
                } else {
                    mprisControlBinding.noPlayers.visibility = View.GONE
                    mprisControlBinding.playerSpinner.visibility = View.VISIBLE
                }

                mprisControlBinding.playerSpinner.onItemSelectedListener =
                    object : AdapterView.OnItemSelectedListener {
                        override fun onItemSelected(arg0: AdapterView<*>?, arg1: View?, pos: Int, id: Long) {
                            if (pos >= playerList.size) return

                            val player = playerList[pos]
                            if (targetPlayer != null && player == targetPlayer!!.playerName) {
                                return  //Player hasn't actually changed
                            }
                            targetPlayer = plugin.getPlayerStatus(player)
                            if (targetPlayer != null) {
                                targetPlayerName = targetPlayer!!.playerName
                            }

                            updatePlayerStatus(plugin)

                            if (targetPlayer != null && targetPlayer!!.isPlaying) {
                                MprisMediaSession.instance.playerSelected(targetPlayer)
                            }
                        }

                        override fun onNothingSelected(arg0: AdapterView<*>?) {
                            targetPlayer = null
                        }
                    }

                if (targetPlayer == null) {
                    //If no player is selected, try to select a playing player
                    targetPlayer = plugin.playingPlayer
                }
                //Try to select the specified player
                if (targetPlayer != null) {
                    val targetIndex = adapter.getPosition(targetPlayer!!.playerName)
                    if (targetIndex >= 0) {
                        mprisControlBinding.playerSpinner.setSelection(targetIndex)
                    } else {
                        targetPlayer = null
                    }
                }
                //If no player selected, select the first one (if any)
                if (targetPlayer == null && playerList.isNotEmpty()) {
                    targetPlayer = plugin.getPlayerStatus(playerList[0])
                    mprisControlBinding.playerSpinner.setSelection(0)
                }
                updatePlayerStatus(plugin)
            }
        }
    }

    private inline fun performActionOnClick(v: View, crossinline l: MprisPlayerCallback) {
        v.setOnClickListener {
            val targetPlayer = targetPlayer ?: return@setOnClickListener
            l(targetPlayer)
        }
    }

    private fun updatePlayerStatus(plugin: MprisPlugin) {
        if (!isAdded) {
            //Fragment is not attached to an activity. We will crash if we try to do anything here.
            return
        }

        var playerStatus = targetPlayer
        if (playerStatus == null) {
            //No player with that name found, just display "empty" data
            playerStatus = plugin.getEmptyPlayer()
        }

        var song = playerStatus.title
        if (!StringUtils.isEmpty(playerStatus.artist)) {
            song += " - " + playerStatus.artist
        }
        if (mprisControlBinding.nowPlayingTextview.text.toString() != song) {
            mprisControlBinding.nowPlayingTextview.text = song
        }

        val albumArt = playerStatus.getAlbumArt()
        if (albumArt == null) {
            val drawable = ContextCompat.getDrawable(requireContext(), R.drawable.ic_album_art_placeholder)!!
            val placeholder_art = DrawableCompat.wrap(drawable)
            activityMprisBinding.albumArt.setImageDrawable(placeholder_art)
        } else {
            activityMprisBinding.albumArt.setImageBitmap(albumArt)
        }

        if (playerStatus.isSeekAllowed) {
            mprisControlBinding.timeTextview.text = durationToProgress(playerStatus.length.milliseconds)
            mprisControlBinding.positionSeek.max = playerStatus.length.toInt()
            mprisControlBinding.positionSeek.progress = playerStatus.position.toInt()
            mprisControlBinding.progressSlider.visibility = View.VISIBLE
        } else {
            mprisControlBinding.progressSlider.visibility = View.GONE
        }

        val volume = playerStatus.volume
        mprisControlBinding.volumeSeek.progress = volume
        if (!playerStatus.isSetVolumeAllowed) {
            mprisControlBinding.volumeSeek.isEnabled = false
        }
        val isPlaying = playerStatus.isPlaying
        if (isPlaying) {
            mprisControlBinding.playButton.setIconResource(R.drawable.ic_pause_black)
            mprisControlBinding.playButton.isEnabled = playerStatus.isPauseAllowed
        } else {
            mprisControlBinding.playButton.setIconResource(R.drawable.ic_play_black)
            mprisControlBinding.playButton.isEnabled = playerStatus.isPlayAllowed
        }

        val loopStatus = playerStatus.loopStatus
        when (loopStatus) {
            "None" -> mprisControlBinding.loopButton.setIconResource(R.drawable.ic_loop_none_black)
            "Track" -> mprisControlBinding.loopButton.setIconResource(R.drawable.ic_loop_track_black)
            "Playlist" -> mprisControlBinding.loopButton.setIconResource(R.drawable.ic_loop_playlist_black)
        }
        val shuffle = playerStatus.shuffle
        if (shuffle) {
            mprisControlBinding.shuffleButton.setIconResource(R.drawable.ic_shuffle_on_black)
        } else {
            mprisControlBinding.shuffleButton.setIconResource(R.drawable.ic_shuffle_off_black)
        }

        mprisControlBinding.loopButton.visibility = if (playerStatus.isLoopStatusAllowed) View.VISIBLE else View.GONE
        mprisControlBinding.shuffleButton.visibility = if (playerStatus.isShuffleAllowed) View.VISIBLE else View.GONE
        mprisControlBinding.volumeLayout.visibility =
            if (playerStatus.isSetVolumeAllowed) View.VISIBLE else View.GONE
        mprisControlBinding.rewButton.visibility = if (playerStatus.isSeekAllowed) View.VISIBLE else View.GONE
        mprisControlBinding.ffButton.visibility =
            if (playerStatus.isSeekAllowed) View.VISIBLE else View.GONE

        requireActivity().invalidateOptionsMenu()

        //Show and hide previous/next buttons simultaneously
        if (playerStatus.isGoPreviousAllowed || playerStatus.isGoNextAllowed) {
            mprisControlBinding.prevButton.visibility = View.VISIBLE
            mprisControlBinding.prevButton.isEnabled = playerStatus.isGoPreviousAllowed
            mprisControlBinding.nextButton.visibility = View.VISIBLE
            mprisControlBinding.nextButton.isEnabled = playerStatus.isGoNextAllowed
        } else {
            mprisControlBinding.prevButton.visibility = View.GONE
            mprisControlBinding.nextButton.visibility = View.GONE
        }
    }

    /**
     * Change current volume with provided step.
     *
     * @param step step size volume change
     */
    private fun updateVolume(step: Int) {
        if (targetPlayer == null) return

        val newVolume = calculateNewVolume(targetPlayer!!.volume, DEFAULT_MAX_VOLUME, step)

        if (targetPlayer!!.volume != newVolume) {
            targetPlayer!!.sendSetVolume(newVolume)
        }
    }

    override fun onVolumeUp() {
        updateVolume(DEFAULT_VOLUME_STEP)
    }

    override fun onVolumeDown() {
        updateVolume(-DEFAULT_VOLUME_STEP)
    }

    override fun onPrepareOptionsMenu(menu: Menu) {
        menu.clear()
        if (targetPlayer != null && "" != targetPlayer!!.url) {
            menu.add(0, MENU_OPEN_URL, Menu.NONE, R.string.mpris_open_url)
        }
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (targetPlayer != null && item.itemId == MENU_OPEN_URL) {
            try {
                val url = VideoUrlsHelper.formatUriWithSeek(targetPlayer!!.url, targetPlayer!!.position).toString()
                val browserIntent = Intent(Intent.ACTION_VIEW, Uri.parse(url))
                startActivity(browserIntent)
                targetPlayer!!.sendPause()
                return true
            } catch (e: MalformedURLException) {
                e.printStackTrace()
                Toast.makeText(requireContext(), getString(R.string.cant_open_url), Toast.LENGTH_LONG).show()
            } catch (e: ActivityNotFoundException) {
                e.printStackTrace()
                Toast.makeText(requireContext(), getString(R.string.cant_open_url), Toast.LENGTH_LONG).show()
            }
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onSaveInstanceState(outState: Bundle) {
        if (targetPlayer != null) {
            outState.putString("targetPlayer", targetPlayerName)
        }
    }

    companion object {
        const val MENU_OPEN_URL: Int = Menu.FIRST
        fun newInstance(deviceId: String?): MprisNowPlayingFragment {
            val mprisNowPlayingFragment = MprisNowPlayingFragment()

            val arguments = Bundle()
            arguments.putString(MprisPlugin.DEVICE_ID_KEY, deviceId)

            mprisNowPlayingFragment.arguments = arguments

            return mprisNowPlayingFragment
        }

        private fun durationToProgress(duration: Duration): String = buildString {
            val length = duration.inWholeSeconds
            var minutes = length / 60
            if (minutes > 60) {
                val hours = minutes / 60
                minutes %= 60
                append(hours)
                append(':')
                if (minutes < 10) append('0')
            }
            append(minutes)
            append(':')
            val seconds = (length % 60)
            if (seconds < 10) append('0') // needed to show length properly (eg 4:05 instead of 4:5)

            append(seconds)
        }
    }
}
